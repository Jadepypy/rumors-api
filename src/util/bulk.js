const BATCH_SIZE = 100;

export default class Bulk {
  constructor(client, batchSize = BATCH_SIZE) {
    this.client = client;
    this.batchSize = batchSize;
    this._operations = [];
    this.actionsCount = 0;
  }

  async push(items, count = 1) {
    this._operations.push(...items);
    this.actionsCount += count;
    if (
      this.actionCounts > this.batchSize ||
      this._operations.length > 3 * this.batchSize
    ) {
      await this.flush();
    }
    return this;
  }

  async flush() {
    if (this._operations.length === 0) {
      return this;
    }

    // Process entires queued up
    const op = this._operations.splice(0);
    this.actionCounts = 0;
    const { body } = await this.client.bulk({
      body: op,
      refresh: 'true',
    });
    if (body.errors) {
      console.error(body.errors);
    } else {
      console.info(`${body.items.length} items processed`);
    }

    return this;
  }
}
