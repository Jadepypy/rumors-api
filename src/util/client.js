import elasticsearch from '@elastic/elasticsearch';

export default new elasticsearch.Client({
  node: process.env.ELASTICSEARCH_URL,
});

// Processes {_id, _version, found, _source: {...}} to
// {id, ..._source}.
//
export function processMeta({
  _id: id,
  _source: source,

  found, // for mget queries
  _score, // for search queries

  sort, // cursor when sorted
}) {
  if (found || _score !== undefined) {
    return { id, ...source, _cursor: sort, _score };
  }
  return null; // not found
}
