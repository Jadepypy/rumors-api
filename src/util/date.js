const dateFormat = /^[0-9]{4}-[0-9]{2}-[0-9]{2}$/;

/**
 * Given a date string, determine if it's a valid date, and return a Date object if it is.

 * @param {string} dateStr        A string representation of a date in YYYY-MM-DD
 * @param {string} [name='Date']  A string used in error message

 * @return {Date}
 */
const convertAndValidateDate = (dateStr, name = 'Date') => {
  if (!dateStr.match(dateFormat)) {
    throw new Error(`${name} must be in the format of YYYY-MM-DD`);
  }
  try {
    const date = new Date(dateStr);
    if (date.toString() !== 'Invalid Date') {
      return date;
    } else {
      throw new Error();
    }
  } catch {
    throw new Error(`${name} must be a valid date in the format of YYYY-MM-DD`);
  }
};

/**
 * Given two date strings and optional maxDuration, determine if it's a valid
   date range.

 * @param {string} startDateStr             A string representation of start date in YYYY-MM-DD.
 * @param {string} endDateStr               A string representation of end date in YYYY-MM-DD.
 * @param {number} [maxDuration]            Optional maximun duration in milliseconds.
   @param {bool}   [allowFutureDate=false]  Whether the date range allows end dates in the future.
 * @return {object}                         Validation result/error
 */
const validateDateRange = (
  startDateStr,
  endDateStr,
  maxDuration,
  allowFutureDate = false
) => {
  let errorMessage, startDate, endDate;

  if (!startDateStr || !endDateStr) {
    errorMessage = 'must include both start date and end date';
  } else {
    try {
      startDate = convertAndValidateDate(startDateStr, 'start date');
      endDate = convertAndValidateDate(endDateStr, 'end date');
    } catch (e) {
      errorMessage = e.message;
    }
    const duration = endDate - startDate;
    if (duration < 0) {
      errorMessage = 'end date cannot be earlier than start date';
    } else if (maxDuration && duration > maxDuration) {
      // TODO: write a formator that uses the most sensible time unit to describe duration.
      errorMessage = `start date and end date cannot be more than ${maxDuration} apart`;
    } else if (!allowFutureDate && endDate - new Date() > 0) {
      errorMessage = 'end date must be no later than today';
    }
  }
  return {
    isValid: !errorMessage,
    error: errorMessage,
    startDate,
    endDate,
  };
};

export { convertAndValidateDate, validateDateRange };
