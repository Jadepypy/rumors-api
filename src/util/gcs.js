import { Storage } from '@google-cloud/storage';

const storage = new Storage({
  credentials: JSON.parse(process.env.GCS_CREDENTIALS || '{}'),
});
const gcsBucket = storage.bucket(process.env.GCS_BUCKET_NAME || 'default');

/**
 * Generate hash for identifying if two files are similar
 *
 * @param {ReadableStream} fileStream
 * @param {string} fileName
 * @param {string} contentType MIME type
 * @returns {string} url
 */
export async function uploadToGCS(fileStream, fileName, contentType) {
  if (!process.env.GCS_BUCKET_NAME) {
    throw new Error('GCS_BUCKET_NAME is not set, cannot upload file.');
  }

  const options = {
    metadata: {
      contentType,
    },
  };

  const file = gcsBucket.file(fileName);
  const url = await new Promise((resolve, reject) => {
    fileStream
      .pipe(file.createWriteStream(options))
      .on('error', function(err) {
        reject(new Error(`[GCS] ${err}`));
      })
      .on('finish', function() {
        // The file upload is complete.
        resolve(
          `https://storage.googleapis.com/${
            process.env.GCS_BUCKET_NAME
          }/${fileName}`
        );
      });
  });

  return url;
}
