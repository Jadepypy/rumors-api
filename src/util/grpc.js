import rollbar from '../rollbarInstance';
import grpc from 'grpc';
const protoLoader = require('@grpc/proto-loader');

const PROTO_PATH = __dirname + '/urlResolver.proto';
const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const urlResolverProto = grpc.loadPackageDefinition(packageDefinition)
  .urlResolver;

const URL_RESOLVER_URL = process.env.URL_RESOLVER_URL || 'localhost:4000';
const client = new urlResolverProto.UrlResolver(
  URL_RESOLVER_URL,
  grpc.credentials.createInsecure()
);

// Receiving stream response from resolver using GRPC
export default urls =>
  new Promise((resolve, reject) => {
    const call = client.ResolveUrl({ urls });
    const responses = [];
    call.on('data', response => {
      responses.push(response);
    });
    call.on('error', err => {
      // eslint-disable-next-line no-console
      console.error('GraphQL operation contains error:', err);
      rollbar.error(
        'GraphQL error',
        {
          body: JSON.stringify({ urls }),
          url: URL_RESOLVER_URL,
        },
        { err }
      );
      reject(err);
    });
    call.on('end', () => resolve(responses));
  });
