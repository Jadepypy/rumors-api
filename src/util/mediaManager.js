import sharp from 'sharp';
import { MediaManager, variants } from '@cofacts/media-manager';

export const IMAGE_PREVIEW = 'webp600w';
export const IMAGE_THUMBNAIL = 'jpg240h';

const mediaManager = new MediaManager({
  bucketName: process.env.GCS_BUCKET_NAME,
  credentialsJSON: process.env.GCS_CREDENTIALS,
  prefix: process.env.GCS_MEDIA_FOLDER,
  getVariantSettings(options) {
    const { type, contentType } = options;
    switch (type) {
      case 'image':
        return [
          variants.original(contentType),
          {
            name: IMAGE_THUMBNAIL,
            contentType: 'image/jpeg',
            transform: sharp()
              .resize({ height: 240, withoutEnlargement: true })
              .jpeg({ quality: 60 }),
          },
          {
            name: IMAGE_PREVIEW,
            contentType: 'image/webp',
            transform: sharp()
              .resize({ width: 600, withoutEnlargement: true })
              .webp({ quality: 30 }),
          },
        ];
      default:
        return variants.defaultGetVariantSettings(options);
    }
  },
});

export default mediaManager;
