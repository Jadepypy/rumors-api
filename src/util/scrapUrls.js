import rollbar from 'rollbarInstance';
import urlRegex from 'url-regex';
import DataLoader from 'dataloader';

import gql from './gql';

/**
 * Extracts urls from a string.
 * Fetches the specified url data from cache or through scrapping.
 * Optionally writes the scrapped result to database.
 *
 * @param {string} text
 * @param {object} options
 * @param {object} options.cacheLoader - The dataloader that loads result from `urls`, given a URL.
 *                                       If not given, scrapUrls don't read `urls` cache.
 * @param {boolean} options.noFetch - If true, return null when cache does not hit. Default: false.
 * @param {object} options.client - ElasticSearch.js client instance used to write new results to
 *                                  `urls` index. If not given, scrapUrl don't write `urls` cache.
 * @param {number} options.scrapLimit - Limit the number of the new URLs scrapped from the text.
 *                                 Cached entries are not counted in limit.
 * @return {Promise<ScrapResult[]>} - If cannot scrap, resolves to null
 */
async function scrapUrls(
  text,
  { cacheLoader, client, noFetch = false, scrapLimit = 5 } = {}
) {
  let urls = text.match(urlRegex()) || [];
  if (urls.length === 0) return [];

  urls = removeFBCLIDIfExist(urls);

  const scrapLoader = new DataLoader(urls =>
    gql`
      query($urls: [String]!) {
        resolvedUrls(urls: $urls) {
          url
          canonical
          title
          summary
          topImageUrl
          html
          status
          error
        }
      }
    `({ urls }).then(({ data }) => data.resolvedUrls)
  );

  const result = await Promise.all(
    // 1st pass: resolve cache
    urls.map(async url => {
      if (cacheLoader) {
        const result = await cacheLoader.load(url);

        if (result)
          return {
            ...result,
            url, // Overwrite the URL from cache with the actual url in text, because cacheLoader may match canonical URLs
            fromUrlsCache: true,
          };
      }

      return url;
    })
  ).then(results => {
    // 2nd pass: scrap when needed
    let scrappingCount = 0;

    return Promise.all(
      results.map(async result => {
        if (typeof result !== 'string') {
          return result;
        }

        // result is an URL here

        if (noFetch || scrappingCount >= scrapLimit) return null;
        scrappingCount += 1;
        return scrapLoader.load(result);
      })
    );
  });

  if (!client) {
    // client not specified, don't to write to urls
    return result;
  }

  const fetchedAt = new Date();

  // Filter out null, write to "urls" index
  const urlsBody = result.reduce((body, r) => {
    if (!r || r.fromUrlsCache) {
      return body;
    }

    const {
      url,
      canonical,
      title,
      summary,
      topImageUrl,
      html,
      status,
      error,
    } = r;

    return body.concat([
      { index: { _index: 'urls', _type: 'doc' } },
      {
        canonical,
        title,
        summary,
        topImageUrl,
        html,
        url,
        status,
        error,
        fetchedAt,
      },
    ]);
  }, []);

  if (urlsBody.length === 0) {
    // Nothing to write to urls
    return result;
  }

  const insertResult = await client.bulk({
    body: urlsBody,
  });

  if (insertResult.errors) {
    rollbar.error('Urls insert error', insertResult);
  }

  return result;
}

/**
 * Remove fbclid if it exist, handling the url comes from FB
 *
 * @param {string[]} inputTexts - raw urls
 * @return {string[]} - urls without fbclid query parameter
 */
export function removeFBCLIDIfExist(inputTexts) {
  return inputTexts.map(text => {
    const myURL = new URL(text);
    myURL.searchParams.delete('fbclid');
    return myURL.toString();
  });
}

export default scrapUrls;
