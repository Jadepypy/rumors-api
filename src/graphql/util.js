import {
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLString,
  GraphQLInt,
  GraphQLList,
  GraphQLEnumType,
  GraphQLFloat,
  GraphQLNonNull,
  GraphQLID,
  GraphQLBoolean,
} from 'graphql';

import Connection from './interfaces/Connection';
import Edge from './interfaces/Edge';
import PageInfo from './interfaces/PageInfo';
import Highlights from './models/Highlights';
import client from 'util/client';
import { imageHash } from 'image-hash';

// https://www.graph.cool/docs/tutorials/designing-powerful-apis-with-graphql-query-parameters-aing7uech3
//
// Filtering args definition & parsing
//

/**
 * @param {string} typeName
 * @param {GraphQLScalarType} argType
 * @param {string} description
 * @returns {GraphQLInputObjectType}
 */
function getArithmeticExpressionType(typeName, argType, description) {
  return new GraphQLInputObjectType({
    name: typeName,
    description,
    fields: {
      LT: { type: argType },
      LTE: { type: argType },
      GT: { type: argType },
      GTE: { type: argType },
      EQ: { type: argType },
    },
  });
}

export const timeRangeInput = getArithmeticExpressionType(
  'TimeRangeInput',
  GraphQLString,
  'List only the entries that were created between the specific time range. ' +
    'The time range value is in elasticsearch date format (https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-date-format.html)'
);
export const intRangeInput = getArithmeticExpressionType(
  'RangeInput',
  GraphQLInt,
  'List only the entries whose field match the criteria.'
);

/**
 * @param {object} arithmeticFilterObj - {LT, LTE, GT, GTE, EQ}, the structure returned by getArithmeticExpressionType
 * @returns {object} Elasticsearch range filter param
 * @see https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html#range-query-field-params
 */
export function getRangeFieldParamFromArithmeticExpression(
  arithmeticFilterObj
) {
  // EQ overrides all other operators
  if (typeof arithmeticFilterObj.EQ !== 'undefined') {
    return {
      gte: arithmeticFilterObj.EQ,
      lte: arithmeticFilterObj.EQ,
    };
  }

  const conditionEntries = Object.entries(arithmeticFilterObj);

  if (conditionEntries.length === 0) throw new Error('Invalid Expression!');

  return Object.fromEntries(
    conditionEntries.map(([key, value]) => [key.toLowerCase(), value])
  );
}

export const moreLikeThisInput = new GraphQLInputObjectType({
  name: 'MoreLikeThisInput',
  description:
    'Parameters for Elasticsearch more_like_this query.\n' +
    'See: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-mlt-query.html',
  fields: {
    like: {
      type: GraphQLString,
      description: 'The text string to search for.',
    },
    minimumShouldMatch: {
      type: GraphQLString,
      description:
        'more_like_this query\'s "minimum_should_match" query param.\n' +
        'See https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-minimum-should-match.html for possible values.',
    },
  },
});

export function createFilterType(typeName, args) {
  const filterType = new GraphQLInputObjectType({
    name: typeName,
    fields: () => ({
      ...args,
      // TODO: converting nested AND / OR to elasticsearch
      // AND: { type: new GraphQLList(filterType) },
      // OR: { type: new GraphQLList(filterType) },
    }),
  });
  return filterType;
}

//
// Sort args definition & parsing
//

const SortOrderEnum = new GraphQLEnumType({
  name: 'SortOrderEnum',
  values: {
    ASC: { value: 'asc' },
    DESC: { value: 'desc' },
  },
});

/**
 * @param {string} typeName
 * @param {Array<string|{name: string, description: string}>} filterableFields
 * @returns {GraphQLList<GarphQLInputObjectType>} sort input type for an field input argument.
 */
export function createSortType(typeName, filterableFields = []) {
  return new GraphQLList(
    new GraphQLInputObjectType({
      name: typeName,
      description:
        'An entry of orderBy argument. Specifies field name and the sort order. Only one field name is allowd per entry.',
      fields: filterableFields.reduce((fields, field) => {
        const fieldName = typeof field === 'string' ? field : field.name;
        const description =
          typeof field === 'string' ? undefined : field.description;

        return {
          ...fields,
          [fieldName]: { type: SortOrderEnum, description },
        };
      }, {}),
    })
  );
}

export const pagingArgs = {
  first: {
    type: GraphQLInt,
    description: 'Returns only first <first> results',
    defaultValue: 10,
  },
  after: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results after this cursor. cannot be used with "before".',
  },
  before: {
    type: GraphQLString,
    description:
      'Specify a cursor, returns results before this cursor. cannot be used with "after".',
  },
};

/**
 * @param {object[]} orderBy - sort input object type
 * @param {{[string]: (order: object) => object}} fieldFnMap - Defines one elasticsearch sort argument entry for a field
 * @returns {Array<{[string]: {order: string}}>} Elasticsearch sort argument in query body
 */
export function getSortArgs(orderBy, fieldFnMap = {}) {
  return orderBy
    .map(item => {
      const field = Object.keys(item)[0];
      const order = item[field];
      const defaultFieldFn = o => ({ [field]: { order: o } });

      return (fieldFnMap[field] || defaultFieldFn)(order);
    })
    .concat({ _id: { order: 'desc' } }); // enforce at least 1 sort order for pagination
}

// sort: [{fieldName: {order: 'desc'}}, {fieldName2: {order: 'desc'}}, ...]
// This utility function reverts the direction of each sort params.
//
function reverseSortArgs(sort) {
  if (!sort) return undefined;
  return sort.map(item => {
    const field = Object.keys(item)[0];
    const order = item[field].order === 'desc' ? 'asc' : 'desc';
    return {
      [field]: {
        ...item[field],
        order,
      },
    };
  });
}

// Export for custom resolveEdges() and resolveLastCursor()
//
export function getCursor(cursor) {
  return Buffer.from(JSON.stringify(cursor)).toString('base64');
}

export function getSearchAfterFromCursor(cursor) {
  if (!cursor) return undefined;
  return JSON.parse(Buffer.from(cursor, 'base64').toString('utf8'));
}

async function defaultResolveTotalCount({
  first, // eslint-disable-line no-unused-vars
  before, // eslint-disable-line no-unused-vars
  after, // eslint-disable-line no-unused-vars
  ...searchContext
}) {
  return (await client.count({
    ...searchContext,
    body: {
      ...searchContext.body,

      // totalCount cannot support these
      sort: undefined,
      track_scores: undefined,
    },
  })).body.count;
}

export async function defaultResolveEdges(
  { first, before, after, ...searchContext },
  args,
  { loaders }
) {
  if (before && after) {
    throw new Error('Use of before & after is prohibited.');
  }

  const nodes = await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      size: first,
      search_after: getSearchAfterFromCursor(before || after),

      // if "before" is given, reverse the sort order and later reverse back
      //
      sort: before
        ? reverseSortArgs(searchContext.body.sort)
        : searchContext.body.sort,
      highlight: {
        order: 'score',
        fields: {
          text: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
          reference: {
            number_of_fragments: 1, // Return only 1 piece highlight text
            fragment_size: 200, // word count of highlighted fragment
            type: 'plain',
          },
        },
        pre_tags: ['<HIGHLIGHT>'],
        post_tags: ['</HIGHLIGHT>'],
      },
    },
  });

  if (before) {
    nodes.reverse();
  }

  return nodes.map(
    ({ _score: score, highlight, inner_hits, _cursor, ...node }) => ({
      node,
      cursor: getCursor(_cursor),
      score,
      highlight,
      inner_hits,
    })
  );
}

async function defaultResolveLastCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const lastNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    body: {
      ...searchContext.body,
      sort: reverseSortArgs(searchContext.body.sort),
    },
    size: 1,
  }))[0];

  return lastNode && getCursor(lastNode._cursor);
}

async function defaultResolveFirstCursor(
  {
    first, // eslint-disable-line no-unused-vars
    before, // eslint-disable-line no-unused-vars
    after, // eslint-disable-line no-unused-vars
    ...searchContext
  },
  args,
  { loaders }
) {
  const firstNode = (await loaders.searchResultLoader.load({
    ...searchContext,
    size: 1,
  }))[0];

  return firstNode && getCursor(firstNode._cursor);
}

async function defaultResolveHighlights(edge) {
  const { highlight: { text, reference } = {}, inner_hits } = edge;

  const hyperlinks = inner_hits?.hyperlinks.hits.hits?.map(
    ({
      _source: { url },
      highlight: {
        'hyperlinks.title': title,
        'hyperlinks.summary': summary,
      } = {},
    }) => ({
      url,
      title: title ? title[0] : undefined,
      summary: summary ? summary[0] : undefined,
    })
  );

  // Elasticsearch highlight returns an array because it can be multiple fragments,
  // We directly returns first element(text, title, summary) here because we set number_of_fragments to 1.
  return {
    text: text ? text[0] : undefined,
    reference: reference ? reference[0] : undefined,
    hyperlinks,
  };
}

// All search
//
export function createConnectionType(
  typeName,
  nodeType,
  {
    // Default resolvers
    resolveTotalCount = defaultResolveTotalCount,
    resolveEdges = defaultResolveEdges,
    resolveLastCursor = defaultResolveLastCursor,
    resolveFirstCursor = defaultResolveFirstCursor,
    resolveHighlights = defaultResolveHighlights,
  } = {}
) {
  return new GraphQLObjectType({
    name: typeName,
    interfaces: [Connection],
    fields: () => ({
      totalCount: {
        type: new GraphQLNonNull(GraphQLInt),
        description:
          'The total count of the entire collection, regardless of "before", "after".',
        resolve: resolveTotalCount,
      },
      edges: {
        type: new GraphQLNonNull(
          new GraphQLList(
            new GraphQLNonNull(
              new GraphQLObjectType({
                name: `${typeName}Edge`,
                interfaces: [Edge],
                fields: {
                  node: { type: new GraphQLNonNull(nodeType) },
                  cursor: { type: new GraphQLNonNull(GraphQLString) },
                  score: { type: GraphQLFloat },
                  highlight: {
                    type: Highlights,
                    resolve: resolveHighlights,
                  },
                },
              })
            )
          )
        ),
        resolve: resolveEdges,
      },
      pageInfo: {
        type: new GraphQLNonNull(
          new GraphQLObjectType({
            name: `${typeName}PageInfo`,
            interfaces: [PageInfo],
            fields: {
              lastCursor: {
                type: GraphQLString,
                resolve: resolveLastCursor,
              },
              firstCursor: {
                type: GraphQLString,
                resolve: resolveFirstCursor,
              },
            },
          })
        ),
        resolve: params => params,
      },
    }),
  });
}

/**
 * @param {{status: T}[]} entriesWithStatus - list of objects with "status" field
 * @param {T[]} statuses - list of status to keep
 * @returns {Object[]}
 */
export function filterByStatuses(entriesWithStatus, statuses) {
  return entriesWithStatus.filter(({ status }) => statuses.includes(status));
}

/**
 * @param {Buffer} fileBuffer
 * @param {ArticleTypeEnum} type The article type
 * @returns {string} The hash for identifying if two files are similar
 */
export async function getMediaFileHash(fileBuffer, type) {
  let hash = '';
  if (type === 'IMAGE') {
    hash = await new Promise((resolve, reject) => {
      imageHash({ data: fileBuffer }, 16, true, (error, data) => {
        if (error) {
          console.error(error);
          reject(error);
        } else {
          resolve(data);
        }
      });
    });
  }

  return hash;
}

export const DEFAULT_ARTICLE_REPLY_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_CATEGORY_STATUSES = ['NORMAL'];
export const DEFAULT_REPLY_REQUEST_STATUSES = ['NORMAL'];
export const DEFAULT_ARTICLE_REPLY_FEEDBACK_STATUSES = ['NORMAL'];

/**
 * @param {string} pluralEntityName - the name to display on argument description
 * @returns {object} GraphQL args for common list filters
 */
export function createCommonListFilter(pluralEntityName) {
  return {
    appId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by a specific app.`,
    },
    userId: {
      type: GraphQLString,
      description: `Show only ${pluralEntityName} created by the specific user.`,
    },
    createdAt: {
      type: timeRangeInput,
      description: `List only the ${pluralEntityName} that were created between the specific time range.`,
    },
    ids: {
      type: new GraphQLList(new GraphQLNonNull(GraphQLID)),
      description: `If given, only list out ${pluralEntityName} with specific IDs`,
    },
    selfOnly: {
      type: GraphQLBoolean,
      description: `Only list the ${pluralEntityName} created by the currently logged in user`,
    },
  };
}

/**
 * Attach (mutates) filterQueries with Elasticsearch query objects by args.filter in GraphQL resolver
 *
 * @param {Array<Object>} filterQueries - list of filter queries of Elasticsearch bool query
 * @param {object} filter - args.filter in resolver
 * @param {string} userId - userId for the currently logged in user
 * @param {string} appid - appId for the currently logged in user
 */
export function attachCommonListFilter(filterQueries, filter, userId, appId) {
  ['userId', 'appId'].forEach(field => {
    if (!filter[field]) return;
    filterQueries.push({ term: { [field]: filter[field] } });
  });

  if (filter.createdAt) {
    filterQueries.push({
      range: {
        createdAt: getRangeFieldParamFromArithmeticExpression(filter.createdAt),
      },
    });
  }

  if (filter.ids) {
    filterQueries.push({ ids: { values: filter.ids } });
  }

  if (filter.selfOnly) {
    if (!userId) throw new Error('selfOnly can be set only after log in');
    filterQueries.push({ term: { userId } }, { term: { appId } });
  }
}
