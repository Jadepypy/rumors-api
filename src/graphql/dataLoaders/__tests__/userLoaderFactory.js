import { loadFixtures, unloadFixtures } from 'util/fixtures';
import DataLoaders from '../index';
import fixtures from '../__fixtures__/userLoaderFactory';

const loader = new DataLoaders();

describe('userLoaderFactory', () => {
  beforeAll(() => loadFixtures(fixtures));

  it('should return correct user with given slug', async () => {
    expect(
      await loader.userLoader.load({
        slug: 'abc123',
      })
    ).toMatchSnapshot();
    expect(
      await loader.userLoader.load({
        slug: 'def456',
      })
    ).toMatchSnapshot();
  });

  it('should return null if slug does not exist', async () => {
    expect(
      await loader.userLoader.load({
        slug: 'asdf',
      })
    ).toBe(null);
  });

  afterAll(() => unloadFixtures(fixtures));
});
