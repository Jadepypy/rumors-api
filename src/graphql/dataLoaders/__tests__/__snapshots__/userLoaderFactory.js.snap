// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userLoaderFactory should return correct user with given slug 1`] = `
Object {
  "_cursor": undefined,
  "_fields": undefined,
  "_score": 0.6931472,
  "email": "secret@secret.com",
  "highlight": undefined,
  "id": "test-user",
  "inner_hits": undefined,
  "name": "test user",
  "slug": "abc123",
}
`;

exports[`userLoaderFactory should return correct user with given slug 2`] = `
Object {
  "_cursor": undefined,
  "_fields": undefined,
  "_score": 0.6931472,
  "email": "hi@me.com",
  "highlight": undefined,
  "id": "test-user2",
  "inner_hits": undefined,
  "name": "test user2",
  "slug": "def456",
}
`;
